endpoints/put.go

```diff
@@ -90,7 +90,21 @@ func NewPutHandler(backend backends.Backend, maxNumValues int) func(http.Respons
 			resps.Responses[i].UUID = uuid.NewV4().String()
 			ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
 			defer cancel()
-			err = backend.Put(ctx, resps.Responses[i].UUID, toCache, p.TTLSeconds)
+			// Only allow setting a provided key if configured (and ensure a key is provided).
+			if allowKeys && len(p.Key) > 0 {
+				s, err := backend.Get(ctx, p.Key)
+				if err != nil || len(s) == 0 {
```

Comment:
This does not seem to be a very precise check that the key is not in cache.  It seems an error could indicate some other situation as well.  Maybe it's okay if the subsequent Put operation will handle all scenarios gracefully.  It's hard to guess, not knowing what all the failure scenarios are for the Get.  Is it possible to have a better check here?

Reply:
I checked the GET code and we will report back to the client that the key does not exist if `err != nil`, so it is consistent with that. I can see three possible failure modes that will generate a non-nil error
1) The key really does not exist. All good in this case.
2) The key store has some sort of corruption and the value cannot be retrieved. This should be rare, and in this case the key has effectively become unused again. It is unlikely that an attacker would be persistent enough to continually try to overwrite keys in hopes that this sort of corruption will eventually happen and they can inject a malicious payload that ends up winning the auction.
3) A temporary backend issue where the original key will be available again later. This should also be rare (not as rare as (2)), but if this is the case then the PUT that follows will also likely fail and thus not be an issue.

It would be a lot of work on each individual backend to disentangle the possible error states, and the chance of issue is slim. If we start seeing anything suspect, we can enter a second PR to improve the backend code to improve the error characterization. But as it stands I don't think the effort will be worth the reward.

Reply:
I was also imagining an invalid key value (e.g., string too long/short or with disallowed characters) might lead to an error here though, as I said, maybe the Put would handle that properly.

Reply:
If the key was not valid (as in not usable) then of course there is nothing stored in it and the put should also fail